text/0099-transaction-extension-version.md

```diff
@@ -0,0 +1,79 @@
+# RFC-0099: Introduce a transaction extension version
+
+|                 |                                                                                             |
+| --------------- | ------------------------------------------------------------------------------------------- |
+| **Start Date**  | 03 July 2024                                                                               |
+| **Description** | Introduce a versioning for transaction extensions.                               |
+| **Authors**     | Bastian Köcher                                                                              |
+
+## Summary
+
+This RFC proposes a change to the extrinsic format to include a transaction extension version.
+
+## Motivation
+
+The extrinsic format supports to be extended with transaction extensions. These transaction extensions are runtime specific and can be different per chain. Each transaction extension can add data to the extrinsic itself or extend the signed payload.
+This means that adding a transaction extension is breaking the chain specific extrinsic format. A recent example was the introduction of the [`CheckMetadatHash`](https://github.com/polkadot-fellows/runtimes/pull/337) to Polkadot and all its system chains.
+As the extension was adding one byte to the extrinsic, it broke a lot of tooling. By introducing an extra version for the transaction extensions it will be possible to introduce changes to these transaction extensions while still being backwards compatible. 
+Based on the version of the transaction extensions, each chain runtime could decode the extrinsic correctly and also create the correct signed payload.
+
+## Stakeholders
+
+- Runtime users
+- Runtime devs
+- Wallet devs
+
+## Explanation
+
+[RFC84](https://github.com/paritytech/polkadot-sdk/issues/2415) introduced the extrinsic format `5`. The idea is to piggyback onto this change of the extrinsic format to add the extra version for the transaction extensions. If required, this could also come 
+as extrinsic format `6`, but `5` is not yet deployed anywhere. 
+
+The extrinsic format supports the following types of transactions:
+- `Bare`: Does not add anything to the extrinsic.
+- `Signed`: `(Address, Signature, Extensions)`
+- `General`: `Extensions`
+
+The `Signed` and `General` transaction would change to:
+
+- `Signed`: `(Address, Signature, Compact, Extensions)`
+- `General`: `(Compact, Extensions)`
```

Comment:
One thing to consider here is that for the General extrinsic we need to pass the extension version to `validate` function of `TransactionExtension` trait (from https://github.com/paritytech/polkadot-sdk/pull/3685). Only this way we can include it to the payload of a signature which most probably (if any) will be placed in the first extension. I think it can be passed as `inherited_implication`, together with a `call`.

Reply:
This RFC is only about changing the extrinsic format and doesn't really care how the implementation is working ;)

Reply:
It does not pass down the version since only the set of TEs is versioned - not the TEs individually.   
So if you want to do a breaking change in a TE, you need to add a new version the transaction format.

Reply:
Yes, I understand it. 

But I think we will need it if we want it to be part of the signature payload for General extrinsics with the extensions like - https://github.com/paritytech/polkadot-sdk/blob/31dcabf0f83c74f891dfe1976bbb89b07c6abdf1/substrate/frame/support/src/transaction_extensions.rs#L86

Reply:
I still don't get why you need to pass it there. If you are talking about an internal change to the extension itself. This will not work with this number. Basically we will then need to copy the code and create a new extension that introduces the change. On the runtime level we then decide which type to call based on the version.

But we can discuss in the implementation.